src/main/java/combat_system/Combat.java

```diff
@@ -0,0 +1,199 @@
+        }
+        else {
+            return "The attack misses!";
+        }
+    }
+
+    /**
+     * This method determines if a character is an NPC or a player and then proceeds to have them take a turn.
+     * @param participant the GameCharacter who is taking the turn
+     */
+    public void take_turn(GameCharacter participant) {
+        Random r = new Random();
+        if(participant instanceof NonPlayerCharacter) {
+            PlayerCharacter target = find_player();
+            if(r.nextBoolean()) {
+                System.out.println(participant.getName()+" enters a defensive stance");
+            }
+            else {
+                System.out.println(participant.getName()+" makes an attack against you!");
+                System.out.println(damage(r.nextInt(20), target, participant));
+            }
+        }
+        else {
+            System.out.println("It is your turn, what do you do?");
+            System.out.println("Attack");
+            System.out.println("Defend");
+            Scanner sc = new Scanner(System.in);
+            boolean valid_input = false;
+            while(!valid_input) {
+                if(sc.nextLine().equals("Attack")) {
+                    HashMap<String, NonPlayerCharacter> npc_names = new HashMap<>();
+                    System.out.print("The remaining enemies are:");
+                    for(NonPlayerCharacter npc : find_alive_npcs()) {
+                        System.out.println(npc.getName());
+                        npc_names.put(npc.getName(), npc);
+                    }
+                    System.out.println("Which will you attack?");
+                    boolean valid_attack = false;
+                    while(!valid_attack) {
+                        String attack_input = sc.nextLine();
+                        if(npc_names.containsKey(attack_input)) {
+                            System.out.println(damage(r.nextInt(20), npc_names.get(attack_input), participant));
+                            valid_attack = true;
+                        }
+                        else {
+                            System.out.println("Please enter a valid name for the attack");
+                        }
+
+                    }
+                    valid_input = true;
+                }
+                if(sc.nextLine().equals("Defend")) {
+                    System.out.println("You enter a defensive stance");
+                    valid_input = true;
+                }
+                else{
+                    System.out.println("Not a valid input, try again");
+                }
+
+            }
+        }
+    }
+
+    /**
+     * This method finds the player character and returns them from amongst participants
+     * @return the player character
+     */
+    public PlayerCharacter find_player() {
+        for(GameCharacter player : this.participants) {
+            if(player instanceof PlayerCharacter) { // Is this needed?
```

Comment:
you can also do an assert instead of if. Don't know which is better though.

Reply:
I just looked it up and supposedly assert should only be used in testing, but I don't think there is really a big difference. Maybe I will keep it as this just to be safe.